rust/azure_iot_operations_connector/src/base_connector/managed_azure_device_registry.rs

```diff
@@ -44,6 +44,286 @@ pub enum ClientNotification<T> {
     Created(T),
 }
 
+/// Result of a schema modification attempt
+pub enum SchemaModifyResult {
+    /// Indicates that the schema was reported successfully and status was modified
+    Reported(MessageSchemaReference),
+    /// Indicates that the schema or status were modified
+    NotModified,
+}
+
+/// Represents the result of a network modification
+pub enum ModifyResult {
+    /// Indicates that the modification was reported
+    Reported,
+    /// Indicates that the modification did not occur
+    NotModified,
+}
+
+/// A cloneable status reporter for Device and Endpoint status reporting.
+///
+/// This provides a way to report Device and Endpoint status changes from outside the [`DeviceEndpointClient`].
+#[derive(Clone, Debug)]
+pub struct DeviceEndpointStatusReporter {
+    connector_context: Arc<ConnectorContext>,
+    device_endpoint_status: Arc<tokio::sync::RwLock<DeviceEndpointStatus>>,
+    device_endpoint_specification: Arc<std::sync::RwLock<DeviceSpecification>>,
+    device_endpoint_ref: DeviceEndpointRef,
+}
+
+impl DeviceEndpointStatusReporter {
+    /// Used to conditionally report the device status and then updates the device with the new status returned.
+    ///
+    /// The `modify` function is called with the current device status (if any) and should return:
+    /// - `Some(new_status)` if the status should be updated and reported
+    /// - `None` if no update is needed
+    ///
+    /// # Returns
+    /// - [`ModifyResult::Reported`] if the status was updated and successfully reported
+    /// - [`ModifyResult::NotModified`] if no modification was needed or the version changed during processing
+    ///
+    /// # Errors
+    /// [`azure_device_registry::Error`] of kind [`AIOProtocolError`](azure_device_registry::ErrorKind::AIOProtocolError) if
+    /// there are any underlying errors from the AIO RPC protocol. This error will be retried
+    /// 10 times with exponential backoff and jitter and only returned if it still is failing.
+    ///
+    /// [`azure_device_registry::Error`] of kind [`ServiceError`](azure_device_registry::ErrorKind::ServiceError) if an error is returned
+    /// by the Azure Device Registry service.
+    ///
+    /// # Panics
+    /// if the specification mutex has been poisoned, which should not be possible
+    pub async fn report_device_status_if_modified<F>(
+        &self,
+        modify: F,
+    ) -> Result<ModifyResult, azure_device_registry::Error>
+    where
+        F: Fn(Option<Result<(), &AdrConfigError>>) -> Option<Result<(), AdrConfigError>>,
```

Comment:
Since we sometimes call modify twice, should modify return a Cow so that it doesn't need to be cloned twice?